library/char-stream.lisp

```diff
@@ -0,0 +1,427 @@
+
+   ;; common lisp condition classes
+   #:coalton-stream-error
+   #:coalton-stream-error-simple
+   #:coalton-stream-error-simple-message
+   #:coalton-stream-error-closed
+   #:coalton-stream-error-end-of-file
+   #:coalton-stream-error-decoding
+   #:coalton-stream-error-encoding
+   #:coalton-stream-error-timeout
+   #:coalton-stream-error-reader))
+
+#+coalton-release
+(cl:declaim #.coalton-impl:*coalton-optimize-library*)
+
+(cl:in-package #:coalton-library/char-stream)
+
+(cl:defmacro %define-into-as-unsafe-coerce (from-type to-type)
+  (alexandria:with-gensyms (instance)
+    `(define-instance (Into ,from-type ,to-type)
+       (define (into ,instance)
+         (lisp ,to-type (,instance)
+           ,instance)))))
+
+(coalton-toplevel
+  ;; users should interact with streams via the classes `Input' and `Output', rather than explicitly
+  ;; referencing these types
+  (repr :native cl:stream)
+  (define-type %AbstractStream)
+  
+  (repr :native (cl:and cl:stream (cl:satisfies cl:input-stream-p)))
+  (define-type %InputStream)
+
+  (%define-into-as-unsafe-coerce %InputStream %AbstractStream)
+
+  (repr :native (cl:and cl:stream (cl:satisfies cl:output-stream-p)))
+  (define-type %OutputStream)
+
+  (%define-into-as-unsafe-coerce %OutputStream %AbstractStream)
+
+  (repr :native (cl:and cl:stream (cl:satisfies cl:input-stream-p) (cl:satisfies cl:output-stream-p)))
+  (define-type %TwoWayStream)
+
+  (%define-into-as-unsafe-coerce %TwoWayStream %AbstractStream)
+  (%define-into-as-unsafe-coerce %TwoWayStream %InputStream)
+  (%define-into-as-unsafe-coerce %TwoWayStream %OutputStream)
+  
+  ;; variants of this type correspond 1:1 with subclasses of `cl:stream-error' on SBCL
+  (define-type StreamError
+    (StreamErrorSimple String) ; encodes SBCL's `sb-int:simple-stream-error'
+    StreamErrorClosed
+    StreamErrorEndOfFile
+    StreamErrorDecoding
+    StreamErrorEncoding
+    StreamErrorTimeout
+    StreamErrorReader)
+
+  (define-class (Stream :stream)
+    (open? (:stream -> Boolean))
+    (close! (:stream -> Result StreamError Unit)))
+
+  (define-class (Stream :stream => Input :stream)
+    (read-char! (:stream -> Result StreamError Char))
+    (read-line! (:stream -> Result StreamError String)))
+
+  (define-type FlushOperation
+    FlushBlocking
+    FlushAsync)
+
+  (define-class (Stream :stream => Output :stream)
+    (write-char! (:stream -> Char -> Result StreamError Unit))
+    (write-string! (:stream -> String -> Result StreamError Unit))
+    (flush-output! (FlushOperation -> :stream -> Result StreamError Unit)))
+
+  (declare %abstract-stream-open? (%AbstractStream -> Boolean))
+  (define (%abstract-stream-open? stream)
+    (lisp Boolean (stream)
+      (cl:open-stream-p stream)))
+
+  (declare %abstract-stream-close! (%AbstractStream -> Result StreamError Unit))
+  (define (%abstract-stream-close! stream)
+    (if (%abstract-stream-open? stream)
+        (Ok (lisp Unit (stream)
+              (cl:close stream)))
```

Comment:
This throws an error at runtime because `(cl:close stream)` returns a boolean for success, not a unit.

Reply:
Oops.